File: backend/app/services/image_analysis_service.py
"""
Image Analysis Service for Telegram Bot
Analyzes images using GitHub Models API (GPT-4o vision)
"""

import base64
import json
from typing import Dict, Any, Optional
from loguru import logger  # type: ignore
import httpx  # type: ignore

from app.config import settings


async def analyze_image_from_url(image_url: str, filename: str = "image") -> Dict[str, Any]:
    """
    Analyze an image from a URL using GPT-4o vision model
    
    Args:
        image_url: URL to the image (can be Telegram file URL)
        filename: Original filename for context
    
    Returns:
        Dictionary with title, content, tags, category, and confidence
    """
    try:
        prompt = """Analyze this image and provide the following in JSON format:
1. A concise, descriptive title (max 100 chars)
2. A detailed description of what's in the image (max 500 chars)
3. 3-5 relevant tags/keywords
4. A category from: Personal, Work, Travel, Food, People, Nature, Technology, Art, Education, Health, Finance, Shopping, Entertainment, Other

Respond ONLY with valid JSON in this exact structure:
{
  "title": "concise title here",
  "content": "detailed description here",
  "tags": ["tag1", "tag2", "tag3"],
  "category": "category here",
  "confidence": 0.95
}

Be specific and accurate. The confidence score should be between 0 and 1."""

        async with httpx.AsyncClient(timeout=30.0) as client:
            response = await client.post(
                f"{settings.GITHUB_MODELS_ENDPOINT}/chat/completions",
                headers={
                    "Content-Type": "application/json",
                    "Authorization": f"Bearer {settings.GITHUB_TOKEN}",
                },
                json={
                    "model": "gpt-4o",
                    "messages": [
                        {
                            "role": "user",
                            "content": [
                                {
                                    "type": "text",
                                    "text": prompt,
                                },
                                {
                                    "type": "image_url",
                                    "image_url": {
                                        "url": image_url,
                                    },
                                },
                            ],
                        }
                    ],
                    "max_tokens": 1000,
                    "temperature": 0.7,
                },
            )

        if response.status_code != 200:
            logger.error(f"Vision API error: {response.text}")
            return _fallback_analysis(filename)

        data = response.json()
        assistant_message = data.get("choices", [{}])[0].get("message", {}).get("content")

        if not assistant_message:
            return _fallback_analysis(filename)

        # Parse JSON response
        try:
            # Extract JSON from response (in case wrapped in markdown)
            if "```json" in assistant_message:
                json_str = assistant_message.split("```json")[1].split("```")[0].strip()
            elif "```" in assistant_message:
                json_str = assistant_message.split("```")[1].split("```")[0].strip()
            else:
                json_str = assistant_message.strip()
            
            # Try to find JSON object
            if "{" in json_str and "}" in json_str:
                start = json_str.index("{")
                end = json_str.rindex("}") + 1
                json_str = json_str[start:end]
            
            analysis = json.loads(json_str)
        except json.JSONDecodeError as e:
            logger.error(f"Failed to parse AI response: {e}\nResponse: {assistant_message}")
            return _fallback_analysis(filename, assistant_message)

        # Validate and sanitize
        return {
            "title": _sanitize_text(analysis.get("title", _extract_title_from_filename(filename)), 100),
            "content": _sanitize_text(analysis.get("content", ""), 500),
            "tags": [_sanitize_text(tag, 30) for tag in analysis.get("tags", [])[:5]],
            "category": _sanitize_text(analysis.get("category", "Other"), 50),
            "confidence": max(0, min(1, float(analysis.get("confidence", 0.5)))),
        }

    except Exception as e:
        logger.error(f"Image analysis failed: {e}")
        return _fallback_analysis(filename)


async def analyze_image_from_bytes(
    image_bytes: bytes, 
    mime_type: str = "image/jpeg",
    filename: str = "image"
) -> Dict[str, Any]:
    """
    Analyze an image from raw bytes
    
    Args:
        image_bytes: Raw image bytes
        mime_type: MIME type of the image
        filename: Original filename for context
    
    Returns:
        Dictionary with title, content, tags, category, and confidence
    """
    try:
        # Convert to base64
        base64_image = base64.b64encode(image_bytes).decode('utf-8')
        data_url = f"data:{mime_type};base64,{base64_image}"
        
        # Use the URL-based analysis
        return await analyze_image_from_url(data_url, filename)
    
    except Exception as e:
        logger.error(f"Image bytes analysis failed: {e}")
        return _fallback_analysis(filename)


def _fallback_analysis(filename: str, content: str = "") -> Dict[str, Any]:
    """
    Generate fallback analysis when AI fails
    """
    title = _extract_title_from_filename(filename)
    
    return {
        "title": title,
        "content": content[:500] if content else "Image uploaded from Telegram",
        "tags": [],
        "category": "Other",
        "confidence": 0.0,
    }


def _extract_title_from_filename(filename: str) -> str:
    """
    Extract a reasonable title from filename
    """
    # Remove extension
    name_without_ext = filename.rsplit(".", 1)[0] if "." in filename else filename
    
    # Replace special characters with spaces
    cleaned = name_without_ext.replace("_", " ").replace("-", " ")
    
    # Capitalize first letter of each word
    title = " ".join(word.capitalize() for word in cleaned.split())
    
    return title if title else "Uploaded Image"


def _sanitize_text(text: str, max_length: int) -> str:
    """
    Sanitize and truncate text
    """
    if not text:
        return ""
    
    # Remove excessive whitespace
    cleaned = " ".join(text.split())
    
    # Truncate if needed
    if len(cleaned) > max_length:
        return cleaned[:max_length - 3] + "..."
    
    return cleaned

File: lib/services/imageAnalysisService.ts
/**
 * Image Analysis Service
 * 
 * Modular service for analyzing images using AI vision models
 * Generates titles, descriptions, and metadata for uploaded images
 */

export interface ImageAnalysisResult {
  title: string;
  content: string;
  suggestedTags?: string[];
  suggestedCategory?: string;
  confidence: number;
}

export interface ImageAnalysisOptions {
  generateTitle?: boolean;
  generateContent?: boolean;
  generateTags?: boolean;
  generateCategory?: boolean;
  maxTitleLength?: number;
  maxContentLength?: number;
}

/**
 * Analyzes an image file and generates metadata
 */
export async function analyzeImage(
  file: File,
  options: ImageAnalysisOptions = {}
): Promise<ImageAnalysisResult> {
  const {
    generateTitle = true,
    generateContent = true,
    generateTags = true,
    generateCategory = true,
    maxTitleLength = 100,
    maxContentLength = 500,
  } = options;

  try {
    // Convert file to base64
    const base64Image = await fileToBase64(file);

    // Call the image analysis API
    const response = await fetch('/api/analyze-image', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        image: base64Image,
        filename: file.name,
        mimeType: file.type,
        options: {
          generateTitle,
          generateContent,
          generateTags,
          generateCategory,
          maxTitleLength,
          maxContentLength,
        },
      }),
    });

    if (!response.ok) {
      throw new Error(`Image analysis failed: ${response.statusText}`);
    }

    const result = await response.json();
    return result;
  } catch (error) {
    console.error('Image analysis error:', error);
    
    // Return fallback result based on filename
    const fallbackTitle = extractTitleFromFilename(file.name);
    return {
      title: fallbackTitle,
      content: '',
      confidence: 0,
    };
  }
}

/**
 * Analyzes multiple images in batch
 */
export async function analyzeImages(
  files: File[],
  options: ImageAnalysisOptions = {}
): Promise<ImageAnalysisResult[]> {
  // Analyze images in parallel for better performance
  const promises = files.map(file => analyzeImage(file, options));
  return Promise.all(promises);
}

/**
 * Checks if a file is an image
 */
export function isImageFile(file: File): boolean {
  return file.type.startsWith('image/');
}

/**
 * Converts a File to base64 string
 */
async function fileToBase64(file: File): Promise<string> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.onload = () => {
      const base64 = reader.result as string;
      // Remove data URL prefix (e.g., "data:image/png;base64,")
      const base64Content = base64.split(',')[1];
      resolve(base64Content);
    };
    reader.onerror = reject;
    reader.readAsDataURL(file);
  });
}

/**
 * Extracts a reasonable title from filename as fallback
 */
function extractTitleFromFilename(filename: string): string {
  // Remove extension
  const nameWithoutExt = filename.replace(/\.[^/.]+$/, '');
  
  // Replace special characters with spaces
  const cleaned = nameWithoutExt.replace(/[_-]/g, ' ');
  
  // Capitalize first letter of each word
  const title = cleaned
    .split(' ')
    .map(word => word.charAt(0).toUpperCase() + word.slice(1).toLowerCase())
    .join(' ');
  
  return title || 'Uploaded Image';
}

/**
 * Validates image file before analysis
 */
export function validateImageFile(file: File): { valid: boolean; error?: string } {
  // Check if it's an image
  if (!isImageFile(file)) {
    return { valid: false, error: 'File is not an image' };
  }

  // Check file size (max 10MB for analysis)
  const maxSize = 10 * 1024 * 1024;
  if (file.size > maxSize) {
    return { valid: false, error: 'Image file too large (max 10MB)' };
  }

  // Check supported formats
  const supportedFormats = ['image/jpeg', 'image/jpg', 'image/png', 'image/gif', 'image/webp'];
  if (!supportedFormats.includes(file.type)) {
    return { valid: false, error: 'Unsupported image format. Please use JPEG, PNG, GIF, or WebP' };
  }

  return { valid: true };
}
